tests/WeakEvent.Tests/WeakEventSourceAsyncTests.cs

```diff
@@ -0,0 +1,307 @@
+
+            await pub.Raise();
+            calledSubscribers.Should().Equal(1);
+
+            // Make sure sub1 is not collected before the end of the test (sub2 can be collected)
+            GC.KeepAlive(sub1);
+        }
+
+        [Fact]
+        public async Task Reentrant_Subscribers_Dont_Fire_Immediately()
+        {
+            var pub = new Publisher();
+            var calledSubscribers = new List<int>();
+            var sub2 = new InstanceSubscriber(2, i => AsyncAdd(calledSubscribers, i));
+            var sub1 = new InstanceSubscriber(1, async i =>
+            {
+                await AsyncAdd(calledSubscribers, i);
+
+                // This listener should not receive the event during the first round of notifications
+                sub2.Subscribe(pub);
+
+                // Make sure subscribers are not collected before the end of the test
+                GC.KeepAlive(sub2);
+            });
+            sub1.Subscribe(pub);
+
+            await pub.Raise();
+            calledSubscribers.Should().Equal(1);
+
+            // Make sure subscribers are not collected before the end of the test
+            GC.KeepAlive(sub1);
+        }
+
+        [Fact]
+        public async Task Multicast_Handlers_Are_Correctly_Unsubscribed()
+        {
+            var pub = new Publisher();
+            var calledSubscribers = new List<int>();
+            AsyncEventHandler<EventArgs> handler = null;
+            handler += (sender, e) => AsyncAdd(calledSubscribers, 1);
+            handler += (sender, e) => AsyncAdd(calledSubscribers, 2);
+
+            pub.Foo += handler;
+            await pub.Raise();
+
+            pub.Foo -= handler;
+            await pub.Raise();
+
+            calledSubscribers.Should().Equal(1, 2);
+        }
+
+        [Fact]
+        public void Can_Subscribe_While_Another_Thread_Is_Invoking()
+        {
+            var sub1CanFinish = new ManualResetEvent(false);
+            var sub2CanSubscribe = new ManualResetEvent(false);
+
+            var pub = new Publisher();
+            var sub1 = new InstanceSubscriber(1, i =>
+            {
+                sub2CanSubscribe.Set();
+                sub1CanFinish.WaitOne();
+                return Task.CompletedTask;
+            });
+            var sub2 = new InstanceSubscriber(2, i => Task.CompletedTask);
+            sub1.Subscribe(pub);
+
+            var thread1 = new Thread(() =>
+            {
+                pub.Raise().Wait();
+            });
+
+            var thread2 = new Thread(() =>
+            {
+                sub2CanSubscribe.WaitOne();
+                sub2.Subscribe(pub);
+            });
+
+            thread1.Start();
+            thread2.Start();
+            bool subscribeFinished = thread2.Join(500);
+            sub1CanFinish.Set();
+            thread1.Join();
+
+            subscribeFinished.Should().BeTrue();
+
+            GC.KeepAlive(sub1);
+            GC.KeepAlive(sub2);
+        }
```

Comment:
Could you try to rewrite this test in a more async fashion? i.e. Use `Task.Run` rather than `new Thread`, use `await` instead of `.Wait()`. There's no `WaitOneAsync` method on `ManualResetEvent`, but you can add it, as shown [here](https://www.thomaslevesque.com/2015/06/04/async-and-cancellation-support-for-wait-handles/)

Reply:
I kept the manual reset event as-is, because it felt weird to include all that code so that the test looked uniform.

Reply:
Is it ok if I leave it?